File: entities/mainer/src/main/java/com/inquallity/daggersandbox/mainer/Mainer.java
package com.inquallity.daggersandbox.mainer;

import android.util.Log;

import javax.inject.Inject;

public class Mainer implements IMainer {

    @Inject
    public Mainer() {
    }

    private void privateMain() {
        Log.d("OYAEBU", "Do private work; $this");
    }

    @Override
    public void doMain() {
        Log.d("OYAEBU", "Mainer doing main in $this");
    }

    @Override
    public void initialize() {
        Log.d("OYAEBU", "Do initialization of $this");
    }

    @Override
    public void doMain2() {
        privateMain();
    }
}

File: entities/mainer/src/main/java/com/inquallity/daggersandbox/mainer/Mainer.kt
package com.inquallity.daggersandbox.mainer

import android.util.Log
import javax.inject.Inject

// Mainer это Entity - самое дно зависимостей
// Изменяем, но никто это API не использует
// Одинаковое поведение
// Меняем код в Core модуле - средний по зависимостям
class Mainer @Inject constructor() : IMainer {

    override fun doMain() {
        Log.d("OYAEBU", "Mainer doing main in $this")
    }

    override fun initialize() {
        Log.d("OYAEBU", "Do initialization of $this")
    }

    override fun doMain2() {
        privateMain()
    }

    private fun privateMain() {
        Log.d("OYAEBU", "Do private work; $this")
    }
}

interface IMainer {
    fun doMain()
    fun initialize()
    fun doMain2()
}